cuda/test/matrix/csr_kernels2.cpp

```diff
@@ -0,0 +1,139 @@
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include <algorithm>
+#include <numeric>
+#include <random>
+#include <vector>
+
+
+#include <gtest/gtest.h>
+
+
+#include <ginkgo/core/base/exception.hpp>
+#include <ginkgo/core/base/executor.hpp>
+#include <ginkgo/core/matrix/coo.hpp>
+#include <ginkgo/core/matrix/csr.hpp>
+#include <ginkgo/core/matrix/dense.hpp>
+#include <ginkgo/core/matrix/diagonal.hpp>
+#include <ginkgo/core/matrix/identity.hpp>
+#include <ginkgo/core/matrix/sparsity_csr.hpp>
+
+
+#include "core/matrix/csr_kernels.hpp"
+#include "core/test/utils.hpp"
+#include "core/test/utils/unsort_matrix.hpp"
+
+
+namespace {
+
+
+class Csr : public ::testing::Test {
+protected:
+    using Arr = gko::Array<int>;
+    using Mtx = gko::matrix::Csr<>;
+    using Vec = gko::matrix::Dense<>;
+
+    Csr()
+#ifdef GINKGO_FAST_TESTS
+        : mtx_size(152, 185),
+#else
+        : mtx_size(532, 231),
+#endif
+          rand_engine(42)
+    {}
+
+    void SetUp()
+    {
+        ref = gko::ReferenceExecutor::create();
+        cuda = gko::CudaExecutor::create(0, ref);
+    }
+
+    void TearDown()
+    {
+        if (cuda != nullptr) {
+            ASSERT_NO_THROW(cuda->synchronize());
+        }
+    }
+
+    template <typename MtxType>
+    std::unique_ptr<MtxType> gen_mtx(int num_rows, int num_cols,
+                                     int min_nnz_row, int max_nnz_row)
+    {
+        return gko::test::generate_random_matrix<MtxType>(
+            num_rows, num_cols,
+            std::uniform_int_distribution<>(min_nnz_row, max_nnz_row),
+            std::normal_distribution<>(-1.0, 1.0), rand_engine, ref);
+    }
+
+    template <typename MtxType>
+    std::unique_ptr<MtxType> gen_mtx(int num_rows, int num_cols,
+                                     int min_nnz_row)
+    {
+        return gen_mtx<MtxType>(num_rows, num_cols, min_nnz_row, num_cols);
+    }
+
+    void set_up_apply_data(int num_vectors = 1)
+    {
+        mtx = Mtx::create(ref);
+        mtx->copy_from(gen_mtx<Vec>(mtx_size[0], mtx_size[1], 1));
+        dmtx = Mtx::create(cuda);
+        dmtx->copy_from(mtx.get());
+    }
+
+    std::shared_ptr<gko::ReferenceExecutor> ref;
+    std::shared_ptr<const gko::CudaExecutor> cuda;
+
+    const gko::dim<2> mtx_size;
+    std::ranlux48 rand_engine;
+    std::unique_ptr<Mtx> mtx;
+    std::unique_ptr<Mtx> dmtx;
+};
+
+
+TEST_F(Csr, CreateSubMatrixIsEquivalentToRef)
+{
+    using Mtx = gko::matrix::Csr<>;
+    set_up_apply_data();
+
+    gko::span rspan{36, 98};
+    gko::span cspan{26, 104};
+    auto smat1 = this->mtx->create_submatrix(rspan, cspan);
+    auto sdmat1 = this->dmtx->create_submatrix(rspan, cspan);
+    auto hmat = Mtx::create(this->ref);
+    hmat->copy_from(sdmat1.get());
+
+    GKO_ASSERT_MTX_NEAR(hmat, smat1, 1e-14);
```

Comment:
nit: we copy the values over directly, so this should match bitwise.
```suggestion
    GKO_ASSERT_MTX_NEAR(hmat, smat1, 0);
```